Fix: MXP EXPIRE tag support and custom element attribute handling

mpconley opened the pull request:
#### Brief overview of PR changes/additions

* **Add MXP EXPIRE tag support**: Implements the EXPIRE attribute for MXP SEND tags, allowing servers to expire groups of links by name
* **Tooltip fix**: Fixed MXP link tooltips displaying "HREF" instead of the actual link command when the EXPIRE attribute appears before the HREF attribute in SEND tags
* **Custom element fix**: Fixed custom MXP elements to properly pass through ALL attributes (EXPIRE, HREF, HINT, etc.) to their expanded tag definitions

#### Motivation for adding to Mudlet

**EXPIRE tag support**: The MXP specification defines an EXPIRE attribute that allows servers to tag groups of links and expire them all at once (e.g., `<EXPIRE Exits>` expires all links tagged with `EXPIRE="Exits"`). This is essential for dynamic content like room exits that change as players move around.

**Tooltip fix**: Addresses reviewer feedback in PR #8383. When games send MXP links like `<SEND EXPIRE="Exits" HREF="east">East</SEND>`, the tooltip was incorrectly showing the literal text "HREF" instead of "east" (the command that would be executed), making links confusing for users.

**Custom element fix**: When servers define custom elements like `<!ELEMENT Ex '<SEND>'>` and use them with attributes like `<Ex EXPIRE="Exits" HREF="north">North</Ex>`, those attributes weren't being passed through to the expanded SEND tag. This made custom elements less functional than regular tags. Now all attributes are properly inherited.

#### Other info (issues closed, discussion etc)

- Related to PR #8383 (initial EXPIRE tag implementation discussion)
- Addresses @mfontani's comment about custom elements not passing through EXPIRE attributes
- Added comprehensive unit tests:
  - Tooltip attribute ordering tests in `TMxpSendTagHandlerTest`
  - Custom element attribute pass-through test in `TMxpCustomElementTagHandlerTest`
- All existing MXP tests continue to pass (100% success rate)
- Changes summary: 5 files changed, 101 insertions(+), 2 deletions(-)

mpconley committed:
Add MXP EXPIRE tag support for expiring links

Implements the MXP (Mud eXtension Protocol) EXPIRE tag to allow MUD servers
to automatically deactivate clickable links when they become outdated.

The EXPIRE tag supports two usage patterns:
- Standalone: <EXPIRE name> deactivates all links tagged with that name
- As attribute: <SEND href="..." EXPIRE="name"> tags a link for later expiration

Expired links remain visible in scrollback for historical reference but
become non-functional when clicked.

Implementation details:
- Added TMxpExpireTagHandler for processing standalone EXPIRE tags
- Extended TLinkStore with dual-map expire tracking system
- Modified TMxpSendTagHandler and TMxpLinkTagHandler to support EXPIRE attribute
- Extended TMxpClient interface with expire-aware methods
- Added comprehensive unit test stubs
- Debug output wrapped in DEBUG_MXP_PROCESSING conditional compilation

Tested with StickMUD (stickmud.com:7680) which uses this feature to expire
room exit links when players move to new rooms.

Requested-by: @mfontani
See: https://discord.com/channels/279748146316312576/1393998251786768384/1429244386541178911

mpconley committed:
Update email address to mike.conley@stickmud.com

mpconley committed:
Fix: MXP link tooltips showing 'HREF' instead of actual link value

When EXPIRE attribute appeared before HREF in MXP SEND tags, the tooltip
incorrectly displayed the literal string 'HREF' instead of the href value.

Fixed by adding ATTR_HREF to the exclusion check in extractHint() method.
Added tests to verify both attribute orderings work correctly.

mpconley committed:
Merge branch 'development' into feature/mxp-expire-href-tooltip-fix

mfontani commented:
@mpconley also note that while this might fix the `<SEND EXPIRE=Foo HREF="...">` it's quite possible that defining a custom element won't, or at least it doesn't seem to work for me.

That is, a system can something like `\e[4z<!ELEMENT Ex '<SEND>'>\e[7z` to define a custom `Ex` element to be used for exits, so it's clearer to the user that that "SEND" element is "for" an exit.

Then, they'd want to send something like `\e[4z<EXPIRE Exits>\e[7z` to expire all previously "tagged" elements, followed by one or more things like `\e[4z<Ex EXPIRE="Exits" HREF="north">\e[7zNorth\e[4z</Ex>\e[7z`, which _works_ the same as if a `<SEND EXPIRE="Exits" ...` were used, _but_ the client can "know", if configured to do so, that that thing "is" an exit.

This part doesn't seem to work with the "EXPIRE": it would seem that the fact that such an `Ex` element "is", transitively, "a `SEND` element" is lost.

Works properly as a "click to send command", though!

mpconley commented:
> @mpconley also note that while this might fix the `<SEND EXPIRE=Foo HREF="...">` it's quite possible that defining a custom element won't, or at least it doesn't seem to work for me.

Thanks for catching this @mfontani! You're absolutely right - custom elements weren't passing through attributes like EXPIRE to their expanded tags.

I've just pushed a fix in commit c75054e that ensures all attributes from custom element tags are passed through to the definition tag, not just those referenced in the template.

What changed:

```<!ELEMENT Ex '<SEND>'>``` used as `<Ex EXPIRE="Exits" HREF="north">` now correctly expands to `<SEND EXPIRE="Exits" HREF="north">`
Added test coverage for this scenario
The fix applies to any custom element, not just exits
This should now work exactly as you'd expect! The EXPIRE functionality will work with custom elements that transitively resolve to SEND tags.

mpconley committed:
Merge branch 'development' into feature/mxp-expire-href-tooltip-fix

mpconley commented:
I am not expecting more changes to this PR. While there are more opportunities outside of the scope for this PR, this should do it for the EXPIRE tag and tags that could be related to it through a thorough review of the specification.

mpconley committed:
Fix: Pass through all attributes from custom MXP elements to definition tags

When a custom MXP element expands to a tag definition, all attributes
from the custom element tag should now be passed through to the expanded
tag. This ensures that attributes like EXPIRE and HREF work correctly
with custom elements.

For example:
  <!ELEMENT Ex '<SEND>'>
  <Ex EXPIRE='Exits' HREF='north'>North</Ex>

Now correctly expands to:
  <SEND EXPIRE='Exits' HREF='north'>North</SEND>

This fix enables hover tooltips to work properly when using custom
elements that expand to SEND tags.

Changes:
- Modified TMxpCustomElementTagHandler::resolveElementDefinition() to
  collect attributes from both the transformed definition and the
  custom tag, combining them into the resulting tag
- Added testCustomElementAttributePassThrough() test case to verify
  attribute pass-through behavior
- Updated TMxpStubClient to capture expireName for testing

Addresses comment in PR #8431
